src/python/zquantum/core/cost_function.py

```diff
@@ -0,0 +1,144 @@
+from .interfaces.cost_function import CostFunction
+from .interfaces.backend import QuantumBackend
+from .circuit import build_ansatz_circuit
+from typing import Callable, Optional, Dict
+import numpy as np
+import copy
+from openfermion import SymbolicOperator
+
+class BasicCostFunction(CostFunction):
+    """
+    Basic implementation of the CostFunction interface.
+    It allows to pass any function (and gradient) when initialized.
+
+    Args:
+        function (Callable): function we want to use as our cost function. Should take a numpy array as input and return a single number.
+        gradient_function (Callable): function used to calculate gradients. Optional.
+        gradient_type (str): parameter indicating which type of gradient should be used.
+        save_evaluation_history (bool): flag indicating whether we want to store the history of all the evaluations.
+        epsilon (float): epsilon used for calculating gradient using finite difference method.
+
+    Params:
+        function (Callable): see Args
+        gradient_function (Callable): see Args
+        evaluations_history (list): List of the tuples (parameters, value) representing all the evaluation in a chronological order.
+        gradient_type (str): see Args
+        save_evaluation_history (bool): see Args
+        epsilon (float): see Args
+
+    """
+
+    def __init__(self, function:Callable, 
+                        gradient_function:Optional[Callable]=None, 
+                        gradient_type:str='custom',
```

Comment:
How come we are defaulting to `custom` instead of `finite_difference`?

Reply:
If user provides the `gradient_function`, they would expect this function to be used. So we default to `custom` and fall back to `finite_difference` if `custom` is not provided.
I think this way we'll cause less harm if someone mess it up.